File: bindings/elixir/examples/matrix_operations.exs
#!/usr/bin/env elixir

# Matrix Operations Example for UMICP Elixir SDK

Code.prepend_path("_build/dev/lib/umicp/ebin")

defmodule MatrixOperationsExample do
  alias UMICP.Matrix.Operations

  def run do
    IO.puts("\n=== Matrix Operations Example ===\n")

    # Vector addition
    IO.puts("Vector Addition:")
    v1 = [1.0, 2.0, 3.0, 4.0]
    v2 = [5.0, 6.0, 7.0, 8.0]
    {:ok, sum} = Operations.vector_add(v1, v2)
    IO.puts("  #{inspect(v1)} + #{inspect(v2)} = #{inspect(sum)}")

    # Vector subtraction
    IO.puts("\nVector Subtraction:")
    {:ok, diff} = Operations.vector_subtract(v2, v1)
    IO.puts("  #{inspect(v2)} - #{inspect(v1)} = #{inspect(diff)}")

    # Scalar multiplication
    IO.puts("\nScalar Multiplication:")
    scaled = Operations.vector_scale(v1, 2.0)
    IO.puts("  #{inspect(v1)} * 2.0 = #{inspect(scaled)}")

    # Dot product
    IO.puts("\nDot Product:")
    {:ok, dot} = Operations.dot_product(v1, v2)
    IO.puts("  #{inspect(v1)} · #{inspect(v2)} = #{dot}")

    # Vector magnitude
    IO.puts("\nVector Magnitude:")
    test_vector = [3.0, 4.0]
    magnitude = Operations.vector_magnitude(test_vector)
    IO.puts("  ||#{inspect(test_vector)}|| = #{magnitude}")

    # Vector normalization
    IO.puts("\nVector Normalization:")
    {:ok, normalized} = Operations.vector_normalize(test_vector)
    IO.puts("  normalize(#{inspect(test_vector)}) = #{inspect(normalized)}")
    norm_magnitude = Operations.vector_magnitude(normalized)
    IO.puts("  magnitude of normalized = #{norm_magnitude}")

    # Matrix multiplication
    IO.puts("\nMatrix Multiplication (2x2):")
    m1 = [1.0, 2.0, 3.0, 4.0]
    m2 = [5.0, 6.0, 7.0, 8.0]
    IO.puts("  [[1, 2],     [[5, 6],")
    IO.puts("   [3, 4]]  *   [7, 8]]")

    {:ok, result} = Operations.matrix_multiply(m1, m2, 2, 2, 2)
    [r0, r1, r2, r3] = result
    IO.puts("  = [[#{r0}, #{r1}],")
    IO.puts("     [#{r2}, #{r3}]]")

    IO.puts("\n✅ Matrix operations example completed!")
  end
end

MatrixOperationsExample.run()

File: bindings/swift/Examples/MatrixOperations.swift
import Foundation
import UMICP

func matrixOperationsExample() throws {
    print("=== Matrix Operations Example ===\n")

    let ops = UMICP.matrixOps()

    // Vector addition
    print("Vector Addition:")
    let v1 = [1.0, 2.0, 3.0, 4.0]
    let v2 = [5.0, 6.0, 7.0, 8.0]
    let sum = try ops.vectorAdd(v1, v2)
    print("  \(v1) + \(v2) = \(sum)")

    // Vector subtraction
    print("\nVector Subtraction:")
    let diff = try ops.vectorSubtract(v2, v1)
    print("  \(v2) - \(v1) = \(diff)")

    // Scalar multiplication
    print("\nScalar Multiplication:")
    let scaled = ops.vectorScale(v1, scalar: 2.0)
    print("  \(v1) * 2.0 = \(scaled)")

    // Dot product
    print("\nDot Product:")
    let dot = try ops.dotProduct(v1, v2)
    print("  \(v1) · \(v2) = \(dot)")

    // Vector magnitude
    print("\nVector Magnitude:")
    let testVector = [3.0, 4.0]
    let magnitude = ops.vectorMagnitude(testVector)
    print("  ||\(testVector)|| = \(magnitude)")

    // Vector normalization
    print("\nVector Normalization:")
    let normalized = try ops.vectorNormalize(testVector)
    print("  normalize(\(testVector)) = \(normalized)")
    print("  magnitude of normalized = \(ops.vectorMagnitude(normalized))")

    // Matrix multiplication
    print("\nMatrix Multiplication (2x2):")
    let m1 = [1.0, 2.0, 3.0, 4.0]
    let m2 = [5.0, 6.0, 7.0, 8.0]
    print("  [[1, 2],     [[5, 6],")
    print("   [3, 4]]  *   [7, 8]]")

    let result = try ops.matrixMultiply(m1, m2, rows: 2, cols: 2, k: 2)
    print("  = [[%g, %g],", result[0], result[1])
    print("     [%g, %g]]", result[2], result[3])

    // Larger matrix multiplication
    print("\nMatrix Multiplication (3x3):")
    let a = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
    let b = [9.0, 8.0, 7.0, 6.0, 5.0, 4.0, 3.0, 2.0, 1.0]
    let matResult = try ops.matrixMultiply(a, b, rows: 3, cols: 3, k: 3)
    print("  Result: \(matResult)")
}

// Run example
do {
    try matrixOperationsExample()
    print("\n✅ Matrix operations example completed!")
} catch {
    print("❌ Error: \(error)")
}

